adapters/criteo/models.go

```diff
@@ -0,0 +1,334 @@
+						if eid.Source == "criteo.com" && len(eid.Uids) > 0 {
+							criteoUser.UID = &eid.Uids[0].ID
+							break
+						}
+					}
+				}
+			}
+		}
+	}
+
+	if device != nil {
+		deviceType := getDeviceType(device.OS)
+		criteoUser.DeviceIDType = &deviceType
+
+		if device.OS != "" {
+			criteoUser.DeviceOs = &device.OS
+		}
+		if device.IFA != "" {
+			criteoUser.DeviceID = &device.IFA
+		}
+		if device.IP != "" {
+			criteoUser.IP = &device.IP
+		}
+		if device.UA != "" {
+			criteoUser.UA = &device.UA
+		}
+	}
+
+	if regsExt != nil {
+		if regsExt.USPrivacy != "" {
+			criteoUser.UspIab = &regsExt.USPrivacy // CCPA
+		}
+	}
+
+	return criteoUser
+}
+
+type criteoGdprConsent struct {
+	GdprApplies *bool   `json:"gdprapplies,omitempty"`
+	ConsentData *string `json:"consentdata,omitempty"`
+}
+
+func newCriteoGdprConsent(user *openrtb.User, regsExt *openrtb_ext.ExtRegs) *criteoGdprConsent {
+	if user == nil && regsExt == nil {
+		return nil
+	}
+
+	var consent *criteoGdprConsent
+	if user != nil {
+		if user.Ext != nil {
+			var userExt *openrtb_ext.ExtUser
+			json.Unmarshal(user.Ext, &userExt)
+			if userExt.Consent != "" {
+				if consent == nil {
+					consent = &criteoGdprConsent{}
+				}
+				consent.ConsentData = &userExt.Consent
+			}
+		}
+	}
+
+	if regsExt != nil {
+		if regsExt.GDPR != nil {
+			if consent == nil {
+				consent = &criteoGdprConsent{}
+			}
+			gdprApplies := bool((*regsExt.GDPR & 1) == 1)
+			consent.GdprApplies = &gdprApplies
+		}
+	}
+
+	return consent
+}
+
+type criteoRequestSlot struct {
+	ImpID       *string              `json:"impid,omitempty"`
+	ZoneID      *uint                `json:"zoneid,omitempty"`
+	NetworkID   *uint                `json:"networkid,omitempty"`
+	PlacementID *string              `json:"placement,omitempty"`
+	Sizes       []*criteoRequestSize `json:"sizes,omitempty"`
+}
+
+func newCriteoRequestSlots(impressions []openrtb.Imp) ([]*criteoRequestSlot, []error) {
+	var errs []error
+
+	if impressions == nil || len(impressions) == 0 {
+		return nil, []error{&errortypes.BadInput{
+			Message: "Bid request impressions is nil or empty",
+		}}
+	}
+
+	// TODO: Criteo slot should comes either with:
+	//   - `zoneid`
+	// OR
+	//   - `networkid`, `placementid`, `sizes`
+	//
+	// if not, criteo will reject the slot.
+	// Would be nice preventing PBS to send such slots when conditions aren't met
+
+	var criteoSlots = make([]*criteoRequestSlot, uint(len(impressions)))
+
+	for i := 0; i < len(impressions); i++ {
+		criteoSlots[i] = &criteoRequestSlot{}
+
+		if impressions[i].ID != "" {
+			criteoSlots[i].ImpID = &impressions[i].ID
+		}
+
+		if impressions[i].Banner != nil {
+			if impressions[i].Banner.Format != nil {
+				criteoSlots[i].Sizes = make([]*criteoRequestSize, len(impressions[i].Banner.Format))
+				for idx, format := range impressions[i].Banner.Format {
+					// TODO: handle properly uint conversion
+					criteoSlots[i].Sizes[idx] = newCriteoRequestSize(uint(format.W), uint(format.H))
```

Comment:
FYI: All `uint`s in the openrtb objects will become signed integers after the library upgrade is complete.

Reply:
Ack